Add Drift Detection Docs and Update Workspaces Notifications Page Content (#55)
hashicorp/terraform-docs-common: website/docs/cloud-docs/workspaces/settings/drift-detection.mdx

---
page_title: Drift Detection - Workspaces - Terraform Cloud and Terraform Enterprise
description: >-
  Enable automatic jobs to check whether your real-world infrastructure matches your Terraform state file.
---

# Drift Detection

> -> **Note:** Drift detection is in beta. Please [provide feedback](https://forms.gle/BwjgvrKQcLbgqK3f7) about how we can improve. Drift detection is available in the paid **Business** tier. Refer to [Terraform Cloud pricing](https://www.hashicorp.com/products/terraform/pricing) for details.

Reviewer: these feel like 2 separate thoughts a little. maybe we should put in separate callouts? or just better combine them, like `Drift detection is available in the **Business** tier and currently in beta.....`

Author: ok i updated and I think it's a bit better!

> -> **Note:** Drift detection is in beta. Please [provide feedback](https://forms.gle/BwjgvrKQcLbgqK3f7) about how we can improve. Drift detection is available in the paid **Business** tier. Refer to [Terraform Cloud pricing](https://www.hashicorp.com/products/terraform/pricing) for details.

Author: ```suggestion
-> **Note:** Drift detection is in beta and is available in the paid **Business** tier. Refer to [Terraform Cloud pricing](https://www.hashicorp.com/products/terraform/pricing) for details and  [provide feedback](https://forms.gle/BwjgvrKQcLbgqK3f7) about how we can improve this feature.
```

> For Terraform, drift means that your real-world infrastructure no longer matches your Terraform state file. Drift can occur when other members of your organization make changes to your infrastructure outside of the official Terraform workflow. For example, a colleague may modify resource configuration directly in the cloud provider console during a production incident.

Author: gonna apply this with an edit to fix a typo :-) 

```suggestion
Infrastructure drift means that your real-world infrastructure no longer matches the configuration in your Terraform state file. Drift occurs when a user modifies resources outside of the Terraform workflow. For example, a colleague may update resource configuration directly in the cloud provider console during a production incident, changing the resource attributes from those in tracked in your state file.
```

> > **Hands On:** Try our [Manage Resource Drift](https://learn.hashicorp.com/tutorials/terraform/resource-drift) tutorial on HashiCorp Learn to practice detecting and fixing differences between your state file and real infrastructure.

Reviewer: ```suggestion
> **Hands On:** Try our [Manage Resource Drift](https://learn.hashicorp.com/tutorials/terraform/resource-drift) tutorial on HashiCorp Learn to review how to detect and resolve drift.
```

When you enable drift detection in a workspace, Terraform Cloud automatically runs a drift detection assessment for that workspace every 24 hours. This involves checking the current state of your real-world infrastructure against the latest version of the workspace’s state file. This lets you monitor drift in your infrastructure without building a custom solution.

## Requirements

Workspaces must must have the following settings to enable drift detection:
- Terraform version 0.15.4+

> - [Remote execution mode](/cloud-docs/workspaces/settings#execution-mode) for Terraform runs

Reviewer 2: We also support the custom [agent pool execution mode](https://www.terraform.io/cloud-docs/agents).

Author: thank you!!! added this + a link :)

> - [Remote execution mode](/cloud-docs/workspaces/settings#execution-mode) for Terraform runs

Author: ```suggestion
- [Remote execution mode](/cloud-docs/workspaces/settings#execution-mode) or [agent pool execution mode](/cloud-docs/agents#configuring-workspaces-to-use-the-agent) for Terraform runs
```

> Workspaces must must have the following settings to enable drift detection:
> - Terraform version 0.15.4+
> - [Remote execution mode](/cloud-docs/workspaces/settings#execution-mode) or [Agent execution mode](/cloud-docs/agents#configuring-workspaces-to-use-the-agent) for Terraform runs

Reviewer: ```suggestion
Drift detection requires the following settings on your workspace: 
- Terraform version 0.15.4+
- [Remote execution mode](/cloud-docs/workspaces/settings#execution-mode) or [Agent execution mode](/cloud-docs/agents#configuring-workspaces-to-use-the-agent) for Terraform runs
```

## Enable Drift Detection

> All workspaces in an organization that meet the [requirements](#requirements) can use drift detection. To enable drift detection within a workspace:
> 1. Go to the workspace and click **Settings > General**.
> 1. Select **Enable** under **Drift Detection**.
> 1. Click **Save settings**.

Reviewer: ```suggestion
To enable drift detection within a workspace:
1. Verify that the workspace satisfies the [requirements](#requirements).
1. Go to the workspace and click **Settings > General**.
1. Select **Enable** under **Drift Detection**.
1. Click **Save settings**.
```

feels a little obvious that it's available to all workspaces with required settings so i moved it into the list of steps. ignore if this feels weird, i dont feel strongly about this one

> Terraform Cloud starts drift detection immediately if there no active Terraform runs in the workspace. If you enable drift detection during a speculative plan, Terraform Cloud will start drift detection soon after completion. If you enable drift detection during other types of runs, Terraform Cloud will start the first automatic drift detection job in about 24 hours.

Author: ```suggestion
Terraform Cloud starts drift detection immediately if there no active Terraform runs in the workspace. If you enable drift detection during a speculative plan, Terraform Cloud will start drift detection soon after completion. If you enable drift detection during other types of runs, Terraform Cloud will start drift detection in about 24 hours.
```

> Drift detection never interrupts or interferes with runs. Terraform Cloud begins drift detection assessments when at least 24 hours have passed since the last assessment and the workspace does not have a run in progress. You can also start new runs during a drift detection assessment. Starting a new run cancels the current assessment, and Terraform Cloud automatically starts the next one in 24 hours.

Reviewer: ```suggestion
Drift detection never interrupts or interferes with runs. Terraform Cloud triggers a drift detection assessment if at least 24 hours have passed since the last assessment and there are no active runs in the workspace. 

You can start a new run during a drift detection assessment, but it will cancel the current assessment. Terraform Cloud will run the next assessment in 24 hours.
```

> Drift detection runs in the background and does not affect your concurrency limit. Terraform Cloud also monitors and controls drift detection concurrency, so it will not cause issues for large-scale deployments with thousands of workspaces. However, drift detection for organizations with a large number of workspaces may take longer to complete.

Reviewer: not clear in here but maybe implied -- does DD run in all workspaces at the same time?

Author: oh yeah this is a fair question. We run them in batches. I didn't want to go too in weeds, but it's fair to mention that here because we talk about it taking longer. Let me adjust!

## Workspace Drift Status

After you enable drift detection, the workspace displays one of the following drift statuses:

> - **Pending:** Terraform Cloud is performing a drift detection assessment for the workspace or there are no prior drift detection results. For example, a workspace will show this status when you enable drift detection during an active run.

Reviewer: based on the rest of the docs here, im not sure which of these is accurate -- 

based on above, enabling dd during an active run means dd will happen in 24 hours. first sentence says a dd is active in the ws. is pending actually that a dd is scheduled? maybe we should list 2 different examples since there are 2 conditions for this?

- **Drifted:** Terraform Cloud has found that one or more real-world infrastructure resources do not match the workspace’s state file.

> -  **Failed:** Terraform Cloud could not successfully complete the drift detection assessment for this workspace and will retry in 24 hours from the failed assessment.

Reviewer: ```suggestion
-  **Failed:** Terraform Cloud could not successfully complete the drift detection assessment for this workspace and will retry 24 hours after the failed assessment.
```

> - **Checked:** Terraform Cloud successfully completed the drift detection assessment for this workspace and found no drift.

Reviewer 2: We've adjusted these in the UI, as being

- Pending
- Drifted
- Failed
- Checked

Author: thank you!! will fix

> - **Pending:** Terraform Cloud is performing a drift detection assessment for the workspace or there are no prior drift detection results. For example, a workspace will show this status when you enable drift detection during an active run.
> - **Drift:** Terraform Cloud has found that one or more real-world infrastructure resources do not match the workspace’s state file.
> -  **Drift Check Failed:** Terraform Cloud could not successfully complete the drift detection assessment for this workspace and will retry in 24 hours.
> - **Checked:** Terraform Cloud successfully completed the drift detection assessment for this workspace and found no drift.

Author: ```suggestion
- **Pending:** Terraform Cloud is performing a drift detection assessment for the workspace or there are no prior drift detection results. For example, a workspace will show this status when you enable drift detection during an active run.
- **Drifted:** Terraform Cloud has found that one or more real-world infrastructure resources do not match the workspace’s state file.
-  **Failed:** Terraform Cloud could not successfully complete the drift detection assessment for this workspace and will retry in 24 hours.
- **Checked:** Terraform Cloud successfully completed the drift detection assessment for this workspace and found no drift.
```

> - **Checked:** Terraform Cloud successfully completed the drift detection assessment for this workspace and found no drift.

Reviewer: ```suggestion
- **Checked:** Terraform Cloud successfully completed the assessment and did not detect drift.
```

> The workspace’s **Drift** tab contains details about the last drift detection assessment. If the workspace has drifted, Terraform Cloud displays a log file showing how the real-world infrastructure has changed from the latest version of the workspace’s state file.

Reviewer 2: We don't show a log file, but a UI similar to the structured run output to display the changes.

I think we could get away with saying something like "Terraform Cloud displays how the real-world infrastructure has changed..."

Author: gotcha, I'll update this!

> ## Correcting Drift

Reviewer: ```suggestion
## Resolving Drift
```

You can use one of the following approaches to correct workspace drift:

> - **Overwrite Drift:** Queue a new plan to realign your real-world infrastructure with your Terraform configuration.

Reviewer: ```suggestion
- **Overwrite drift:** Queue a new plan to realign your real-world infrastructure with your Terraform configuration.
```

> - **Update Terraform state and configuration to match drift:** Queue a [refresh-only plan](/cloud-docs/run/modes-and-options#refresh-only-mode) to update your Terraform state to match your real-world infrastructure. You can also modify your Terraform configuration to include any new or changed resources. Refer to our [Manage Resource Drift](https://learn.hashicorp.com/tutorials/terraform/resource-drift) tutorial for a detailed example.

Reviewer 2: I think we should add a note here reminding folks that if they use the refresh-only strategy **without updating their configuration**, drift will no longer be detected until they run their next apply (that would revert the accepted changes to their state file).

> **Hands On:** Try our [Manage Resource Drift](https://learn.hashicorp.com/tutorials/terraform/resource-drift) tutorial on HashiCorp Learn to review how to detect and resolve drift.

Drift detection is a workspace-level setting. When enabled, Terraform Cloud will automatically run a drift detection assessment for that workspace every 24 hours. The assessment checks the actual settings of your infrastructure against those tracked in your workspace's state file.

## Requirements

Drift detection requires the following settings on your workspace:
- Terraform version 0.15.4+
- [Remote execution mode](/cloud-docs/workspaces/settings#execution-mode) or [Agent execution mode](/cloud-docs/agents#configuring-workspaces-to-use-the-agent) for Terraform runs

## Enable Drift Detection

To enable drift detection within a workspace:
1. Verify that the workspace satisfies the [requirements](#requirements).
1. Go to the workspace and click **Settings > General**.
1. Select **Enable** under **Drift Detection**.
1. Click **Save settings**.

> 

Reviewer: ```suggestion

### Drift detection assessment scheduling

```

> Terraform Cloud starts drift detection immediately if there no active Terraform runs in the workspace. If you enable drift detection during a speculative plan, Terraform Cloud will start drift detection soon after completion. If you enable drift detection during other types of runs, Terraform Cloud will start drift detection in about 24 hours.

Author: ```suggestion
If there are no active Terraform runs in the workspace, Terraform Cloud starts a drift detection assessment right after you enable the feature. If you enable drift detection during a speculative plan, Terraform Cloud will start drift detection soon after that plan's completion. If you enable drift detection during other types of runs, Terraform Cloud will start the first assessment in about 24 hours.
```

> If there are no active Terraform runs in the workspace, Terraform Cloud starts a drift detection assessment right after you enable the feature. If you enable drift detection during a speculative plan, Terraform Cloud will start drift detection soon after that plan's completion. If you enable drift detection during other types of runs, Terraform Cloud will start the first assessment in about 24 hours.

Reviewer: ```suggestion
If there are no active Terraform runs in the workspace, Terraform Cloud starts a drift detection assessment as soon as you enable the feature. If you enable drift detection during a speculative plan, Terraform Cloud will start drift detection soon after that plan's completion. If you enable drift detection during other types of runs, Terraform Cloud will start the first assessment in about 24 hours.
```

Drift detection never interrupts or interferes with runs. Terraform Cloud triggers a drift detection assessment if at least 24 hours have passed since the last assessment and there are no active runs in the workspace. 

You can start a new run during a drift detection assessment, but it will cancel the current assessment. Terraform Cloud will run the next assessment in 24 hours.

> Drift detection runs in the background and does not affect your concurrency limit. Terraform Cloud also monitors and controls drift detection concurrency, so it will not cause issues for large-scale deployments with thousands of workspaces. However, drift detection for organizations with a large number of workspaces may take longer to complete.

Author: ```suggestion
Drift detection runs in the background and does not affect your concurrency limit. Terraform Cloud also monitors and controls drift detection concurrency, so it will not cause issues for large-scale deployments with thousands of workspaces. However, Terraform Cloud performs drift detection assessments in batches, so drift detection may take longer to complete when you enable it in a large number of workspaces.
```
